### DISCUSSION TITLE
Serverless Functions not being cached?

### DISCUSSION THREAD

**Discussion by sr229** [vercel/vercel#5250] [Help]
I've been reading about [Edge caching for Serverless](https://vercel.com/docs/serverless-functions/edge-caching) - and it seems that it allows us to cache API responses from our serverless API to speed up our responses. Our [API](https://github.com/nimiiiii/Nimi) works basically like this: 

![image](https://user-images.githubusercontent.com/14976516/94914170-16b24f00-04dd-11eb-8268-447aa1fae8f1.png)

We assume that that since the API does not change it's contents often, and remains static - we expect that Vercel's CDN would catch this and would just return the data from the CDN that is already cached instead of having to request the Lambdas to re-query the database for the same data. However based on real data, we found out that Vercel's CDN keeps hitting cache misses:

![image](https://user-images.githubusercontent.com/14976516/94914348-70b31480-04dd-11eb-9472-a25a4cb221bf.png)

Therefore, it tells the lambda always to requery our MongoDB Atlas instance - hence making API requests even slower.

![image](https://user-images.githubusercontent.com/14976516/94914446-9f30ef80-04dd-11eb-94c4-3774bf6b1959.png)

Just to clarify we don't change our API responses quite often as it's based off attribute data from publicly available dataset, we just represented it as a JSON API as part of our testing since we are developing a Web application. Therefore based on this data, I believe that:

- Vercel's CDN is not catching our `s-maxage` and `stale-while-revalidate` headers. It should tell the CDN to cache and regenerate the cache but it's not, so we end up with a diagram like the thing stated above.

- We've made all the optimizations we can on our side, however, caching can only get us far in the lambda side.

Is there any Vercel staff that can explain why we keep hitting cache misses despite having relatively "static" API responses.

**Comment by markozxuu** [comment]
Hi @sr229  👋  I think the problem is in the query to the DB. You may have to do something additional so that the connection to the DB is reused in case it is the same user.

**Reply by sr229** [reply]
Thank you for the response @markozxuu. We're using MongoDB atlas with the Mongoose driver so I don't know what else I should add on the API so we actually have to us the cache.

**Reply by leerob** [reply]
It seems there was never any code provided here - could you share that? Sorry for the slow response.

**Reply by sr229** [reply]
> It seems there was never any code provided here - could you share that? Sorry for the slow response.

Sorry for the late reply, our entire code is [Open Source](https://github.com/nimiiiii/nimi-api).

**Reply by leerob** [reply]
First, I'd make sure you're connecting to Mongo with best practices for connection pooling: https://github.com/vercel/next.js/pull/28350

Second, so if you add cache-control headers here, what happens? https://github.com/nimiiiii/nimi-api/blob/581e7738a1169bcc3ebc1ed014a43efe1415225f/lib/createModel.ts#L30 as mentioned below:

```
res.setHeader("Cache-Control", "max-age=0, s-maxage=1800, stale-while-revalidate")
```

**Comment by valse** [comment]
Hi, this is strange because I'm using cached API too and I could check its age in the CDN.
For example I've added these headers:
`res.setHeader("Cache-Control", "max-age=0, s-maxage=1800, stale-while-revalidate")`
For a 30 minutes of cache

**Reply by tonprince** [reply]
Any way to invalidate cache if document is changed in db?

**Reply by leerob** [reply]
@tonprince Currently, no. This work is in progress though (programmatic purging of the cache).

**Reply by tonprince** [reply]
Hi, is there any update on the programmatic purging of the cache in a serverless function?

**Reply by pkrogel-kws** [reply]
I am interested in a response to this as well. I have a client using WordPress for CMS + Edge Caching at 15 minutes, but I'd much rather have it cached indefinitely, and purge cache when content in WordPress is edited.
